sceptre/cli/__init__.py

```diff
@@ -39,21 +41,46 @@
     help="The formatting style for command output.")
 @click.option("--no-colour", is_flag=True, help="Turn off output colouring.")
 @click.option(
-    "--var", multiple=True, help="A variable to template into config files.")
+    "--var", multiple=True,
+    help="A variable to template into config files.")
 @click.option(
     "--var-file", multiple=True, type=click.File("rb"),
     help="A YAML file of variables to template into config files.")
 @click.option(
-    "--ignore-dependencies", is_flag=True, help="Ignore dependencies when executing command.")
+    "--ignore-dependencies", is_flag=True,
+    help="Ignore dependencies when executing command.")
+@click.option(
+    "--merge-keys", is_flag=True, default=False,
+    help="Merge keys in successive var files not overwrite.")
 @click.pass_context
 @catch_exceptions
 def cli(
-        ctx, debug, directory, output, no_colour, var, var_file, ignore_dependencies
+        ctx, debug, directory, output, no_colour, var, var_file, ignore_dependencies, merge_keys
 ):
     """
     Sceptre is a tool to manage your cloud native infrastructure deployments.
-
     """
+    def deep_merge(source, destination):
```

Comment:
I agree with @ngfgrant's suggestion, `I've had a look through and think the changes made would be better suited in the API rather than the CLI. We really want to keep the very vast majority of logic out of the CLI if possible.`?  Have you made an attempt to move this functionality into the API yet?

Reply:
@zaro0508 I am happy to move the functionality, but I would need some guidance as I have noted above. When I looked at it last time, I could not figure out what had to move and where it had to move to. I can have another look perhaps if I get some time. But if it's too difficult, we would probably just use our fork and not submit the feature at all.

Reply:
The idea is that the CLI should only be responsible for gathering the user variables then it should pass those variables along to config/reader.py for processing.  You should do your processing in reader.py which is the API.

I think i would suggest you leave this code as-is because it already gathers all the user vars and passes it along to reader.py.  You can access the all the[ user vars in reader.py](https://github.com/Sceptre/sceptre/blob/master/sceptre/config/reader.py#L132) and update them there.  Also it looks like [config/strategies.py already has a dict_merge](https://github.com/Sceptre/sceptre/blob/master/sceptre/config/strategies.py#L36) method that you can use to merge the keys.

Reply:
@zaro0508 

> I think i would suggest you leave this code as-is because it already gathers all the user vars and passes it along to reader.py.

Unfortunately, that is not quite true. As you can see [here](https://github.com/Sceptre/sceptre/blob/master/sceptre/cli/__init__.py#L76-L82) on lines 76-82, duplicate dictionary keys in the first `var_file` are deleted in favour of those appearing in the second `var_file`. This behaviour is further illustrated in the tests [here](https://github.com/Sceptre/sceptre/blob/master/tests/test_cli.py#L113-L125).

Moreover, the context object does not allow us to pass in the raw dictionaries from the `var_files`, but has only a single field [`user_variables`](https://github.com/Sceptre/sceptre/blob/master/sceptre/cli/__init__.py#L62), where it is expected that these dictionaries are already merged.

The `ConfigReader` class then initialises the `user_variables` [here](https://github.com/Sceptre/sceptre/blob/master/sceptre/config/reader.py#L132.

This is why I did not think my change here should refactor this code. 

I suppose, what we could do is remove the `user_variables` input to `ConfigReader`, and replace it with a list of `--var-files` and a list of `--vars`. Then, the code in the `cli` function could be moved to a private method of the `ConfigReader` class.

But would this change have other consequences? I am not sure.

Reply:
Actually, this would require quite extensive changes all over the place. It is not only the `ConfigReader` class that expects a merged `user_variables` as an input, but the `SceptreContext` class expects it too. 

This tells me that, without a major redesign of the code base, the logic probably belongs in `sceptre.cli.helpers`.

Reply:
@zaro0508 for what it's worth, this is what it looks like after moving the functionality to `sceptre.cli.helpers`: https://github.com/Sceptre/sceptre/pull/928/commits/64dbd6f5e67e4e26c52ef11f8bab8b24b51ae4e7